### PR TITLE
Add a working minipyro with tests

### LINKED ISSUE #57: Fix funsor.minipyro and add tests
[funsor/minipyro.py](https://github.com/pyro-ppl/funsor/blob/master/funsor/minipyro.py) is currently only a storyboard: it does not function but does serve to demonstrate syntax and computations we'd like to support in funsor.

## Tasks
- [ ] add tests for eager sampling
- [ ] add tests for delayed sampling
- [ ] add tests for `trace`
- [ ] add tests for `replay`
- [ ] add tests for `block`
- [ ] add tests for `pyro.plate`
- [ ] add tests for `log_joint`
- [ ] add tests for `barrier`
- [ ] add tests for `ground`

### PR DESCRIPTION
**Pull request by fritzo**
Resolves #57 
Partly pair coded with @eb8680 

This implements:
- minimal version of minipyro
- tests adapted from Pyro's test_valid_models.py

The minipyro in this PR uses a stub for `Expectation` #109, and remaining issues are blocked either by expectation or by missing patterns for `Joint`. Therefore I think it's safe to merge this frontend and move on to more backend work.

## Tested
- [x] added a test_minipyro.py that runs with multiple backends
- [x] ran `pyro/examples/minipyro.py --backend=funsor` (until blocked by `Expectation`)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -17,6 +17,10 @@ install:
       else
           pip install https://download.pytorch.org/whl/cpu/torch-1.0.1.post2-cp36-cp36m-linux_x86_64.whl;
       fi
+
+    # Remove after public release of https://github.com/pyro-ppl/pyro/pull/1794
+    - pip install https://github.com/pyro-ppl/pyro/archive/dev.zip
```
Comment (fritzo) [inline .travis.yml]: @eb8680 You'll need to pull Pyro dev for the new tests to work.

File: funsor/minipyro.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+        params = [site["value"] for site in param_capture.values()]
+        # Take a step w.r.t. each parameter in params.
+        self.optim(params)
+        # Zero out the gradients so that they don't accumulate.
+        for p in params:
+            p.grad = p.new_zeros(p.shape)
+        return loss.item()
+
+
+# TODO(eb8680) Replace this with funsor.Expectation.
```
Comment (fritzo) [inline funsor/minipyro.py]: Deferred to #109

File: funsor/terms.py [hunk 3]
```diff
@@ -745,7 +745,7 @@ def eager_reduce(self, op, reduced_vars):
             assert isinstance(reduced_vars, frozenset)
             reduced_vars = reduced_vars.intersection(self.inputs) | self.reduced_vars
             return Reduce(op, self.arg, reduced_vars)
-        return super(Reduce, self).reduce(op, reduced_vars)
+        return super(Reduce, self).eager_reduce(op, reduced_vars)
```
Comment (fritzo) [inline funsor/terms.py]: This was causing an infinite loop (obvious in retrospect).

File: funsor/minipyro.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+    return msg
+
+
+# sample is an effectful version of Distribution.sample(...)
+# When any effect handlers are active, it constructs an initial message and calls apply_stack.
+def sample(name, fn, obs=None):
+
+    # if there are no active Messengers, we just draw a sample and return it as expected:
+    if not PYRO_STACK:
+        raise NotImplementedError('Funsor cannot sample')
```
Comment (eb8680) [inline funsor/minipyro.py]: Can we return `fn(value=Variable(name, ...)).sample({name}).point` here?

Comment (fritzo) [reply funsor/minipyro.py]: Let me try to do that...

Comment (eb8680) [reply funsor/minipyro.py]: We should probably also check `if not PYRO_STACK and funsor.interpreter._INTERPRETATION is eager` before doing that

**Pull request merged**